### PR TITLE
SPEC-1339 add FLE corpus test

### PR DESCRIPTION
**Pull request by kevinAlbs**
This adds an exhaustive way of testing encryption and decryption of FLE with a single document where each field is encrypted in a different way by enumerating the matrix with the following axes:
- KMS (local, aws)
- BSON type
- algorithm (random, deterministic)
- method (explicit, automatic)
- key identifier (_id, keyAltName)

The test is also run with local and remote configured schema.

In the process of generating this corpus with the Java driver, this caught multiple bugs:
(CDRIVER-3204, CDRIVER-3205, CDRIVER-3216, JAVA-3335, SERVER-41994).

### CODE REVIEW

File: source/client-side-encryption/tests/README.rst [hunk 1]
```diff
@@ -207,32 +207,6 @@ as follows.
    - Expect a `findOne` on the key vault collection with `_id` set to the returned binary to return a document.
    - Expect that document to have "masterKey.provider" set to "aws"
 
-#. Test explicit encrypt and decrypt with the "local" KMS provider.
```
Comment (kevinAlbs) [inline source/client-side-encryption/tests/README.rst]: These tests are now redundant.

**Comment by ajdavis** [comment]
I think there are 40 missing tests:
```
import json

f = open('source/client-side-encryption/corpus/corpus-schema.json')
schema = json.load(f)['properties']

types = ('double',
         'string',
         'object',
         'array',
         'binData=00',
         'binData=04',
         'undefined',
         'objectId',
         'bool',
         'date',
         'null',
         'regex',
         'dbPointer',
         'javascript',
         'int',
         'long',
         'timestamp',
         'decimal',
         'minKey',
         'maxKey',
         'javascriptWithScope',
         'symbol')

restricted = ['bool', 'null', 'undefined', 'minKey', 'maxKey']

keys = set('_'.join([aws, t, ran, expl])
           for aws in ('aws', 'local')
           for t in types
           for ran in ('rand', 'det') for expl in ('auto', 'explicit')
           if (not (t in restricted and ran == 'det')))

from pprint import pprint

print('%s extra tests:' % (len(set(schema.keys()) - keys),))
pprint(set(schema.keys()) - keys)
print()
print()
missing_keys = keys - set(schema.keys())
print('%s missing tests of %s total:' % (len(missing_keys), len(keys)))
pprint(missing_keys)
```
I don't see any tests of restricted types like null, minKey, etc. with random encryption, or decimal with deterministic encryption, and some others.

**Comment by kevinAlbs** [comment]
I excluded prohibited combinations from the corpus, because they would cause the encryption to fail. Though I think it does make sense to include those prohibitions as a separate test. I'll add that.

**Comment by ajdavis** [comment]
Ah, I had lost track of the restrictions on types, I see now that these are indeed comprehensive.

However, in a separate patch let's test that all prohibited types (single-valued types like null and types that cannot be deterministically compared like Decimal128) are indeed prohibited.

**Comment by ajdavis** [comment]
I've checked corpus.json with a similar script and confirmed that all tests are covered.

File: source/client-side-encryption/tests/README.rst [hunk 2]
```diff
@@ -263,4 +237,75 @@ as follows.
...
+
+4. Load `corpus/corpus.json <corpus/corpus.json>`_ to a variable named ``corpus``. Field names have the following layout: `<kms>_<type>_<algo>_<method>`.
+
+   - ``kms`` is either ``aws`` or ``local``
+   - ``type`` is a BSON type string `names coming from here <https://docs.mongodb.com/manual/reference/operator/query/type/>`_)
+   - ``algo`` is either ``rand`` or ``det`` for random or deterministic encryption
+   - ``method`` is either ``auto`` or ``expl`` for automatic or explicit encryption
+   - ``identifier`` is either ``id`` or ``altname`` for the key identifier
+
+   Iterate over each ``_id`` field of ``corpus``. For each field, if the ``method`` is ``expl``, replace the value with a value explicitly encrypted with ``client_encryption``.
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply source/client-side-encryption/tests/README.rst]: Done.

File: source/client-side-encryption/tests/README.rst [hunk 3]
```diff
@@ -263,4 +237,75 @@ as follows.
...
+      Mng0NCt4ZHVUYUJCa1kxNkVyNUR1QURhZ2h2UzR2d2RrZzh0cFBwM3R6NmdWMDFBMUN3YkQ5aXRRMkhGRGdQV09wOGVNYUMxT2k3NjZKelhaQmRCZGJkTXVyZG9uSjFk
+
+   Configure both objects with `keyVaultNamespace` set to `admin.datakeys`.
+
+4. Load `corpus/corpus.json <corpus/corpus.json>`_ to a variable named ``corpus``. Field names have the following layout: `<kms>_<type>_<algo>_<method>`.
+
+   - ``kms`` is either ``aws`` or ``local``
+   - ``type`` is a BSON type string `names coming from here <https://docs.mongodb.com/manual/reference/operator/query/type/>`_)
+   - ``algo`` is either ``rand`` or ``det`` for random or deterministic encryption
+   - ``method`` is either ``auto`` or ``expl`` for automatic or explicit encryption
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply source/client-side-encryption/tests/README.rst]: Done.

**Comment by kevinAlbs** [comment]
Added the generation script, and added the prohibited combinations for explicit encryption. For automatic, I could not think of a reasonable way to test since you cannot configure a schema with prohibited combinations.

**Comment by kevinAlbs** [comment]
Adding @rozza. Ross, to validate this new test I used Java driver: [repo here](https://github.com/kevinAlbs/JavaCorpusPoC). I was hoping this could be used as the basis for the Java driver test as well. Would you mind reviewing this new test, and refactoring that runner into a prose test when this test gets approved?

**Comment by kevinAlbs** [comment]
I like that idea, updated values to be documents with metadata, though I left the field names as is. I cannot put these documents inside an array since it's prohibited to have a JSON schema nesting encrypted fields inside an array.

**Comment by rozza** [comment]
@kevinAlbs happy once this has been approved to take the POC test runner as the basis for the test :+1:

**Comment by kevinAlbs** [comment]
Thanks for reviewing! Rebasing + merging.

**Pull request merged**